### ISSUE TITLE
Add undo to replace all in project

### ISSUE THREAD

**Issue by hachpai** [atom/find-and-replace#331]
Hello,

I've spent 2 hours to undo a bad replace all in project. An not-undoable operation for the moment.
It's REALLY dangerous :-1: 

Miss this function truly.

thank you

ph

**Comment by samtuke** [comment]
Yeah this is crazy. I just experienced:

> Replaced " with '  41197 times in 961 files

Why does Atom replace in files which aren't open? Why did it replace in multiple files in the first place? I see only a "replace all" button, I'm unsure if the changes were saved across all affected files already (appears so, no blue icons on tabs of open files indicating unsaved changes).

Ouch :(

**Comment by benogle** [comment]
> Why did it replace in multiple files in the first place?

There are two kinds of find and replace UIs: find in buffer, and find in project. Sound like find in buffer was what you wanted, but you used find in project.

**Comment by JacobEvelyn** [comment]
+1, I've run into this several times (including now) and it's a huge pain to fix the mistake.

**Comment by dk253** [comment]
twice this week... the last one was 1350 changes in 58 files.  Woohoo.

**Comment by samtuke** [comment]
On 03/03/15 15:33, dk253 wrote:
> twice this week... the last one was 1350 changes in 58 files. Woohoo.

Heck. I feel your pain.

**Comment by btm1** [comment]
i just had this happen too... also where is there only a replace all when you find in project? I just want to find all instances of something. I could have sworn there used to be just a find button.

**Comment by waywalker** [comment]
Wow. Danger :mouse: :-) Just got bitten by this as well.

**Comment by park9140** [comment]
Perhaps a confirm when replacing in unopened files would be appropriate until we can get undo?

**Comment by andya9** [comment]
:+1:  it would be really really useful

**Comment by graemeworthy** [comment]
+1

**Comment by jb** [comment]
I just did this on my entire `~/.atom` directory when trying to replace something in package file. If I'd accidentally opened my home directory instead (which is quite possible for a newb to do by accident), I likely would have lost data.

IMO, this feature should be removed entirely until some safeguards can be put in place, such as:

  - A warning that all files in directory X will be modified
  - A warning that appears if some of the files to be modified are not version-controlled
  - A warning that appears if the project directory is `/`, `/home/*/`, or `/User/*/`
  - An undo feature (that will still work if Atom crashes during the replace)

Because this feature is dangerous (or at the very least a major <abbr title="Pain In The Ass">PITA</abbr>) when used incorrectly, please consider changing the <kbd>enchancement</kbd> label.

@benogle

**Comment by andya9** [comment]
Maybe the most straightforward temporary way could be to remove the find-in-project keybinding from defaults... I think that completely removing the ability to mass replace would completely prevent atom from spreading now that we are next to 1.0...

**Comment by jb** [comment]
Good idea. Although it still doesn't behave how you would expect it to when you open your home directory, then browse to your project. It's not `find-in-project`, it's `find-and-replace-everything-atom-currently-has-access-to`. I really think it should display a warning when you try to replace in a directory that is not version controlled. Just something along the lines of:

    Make replacement in every file in `/User/john/`?

By the way, I just realized that I ended up making this mistake because Vim-mode rebinds <kbd>ctrl</kbd><kbd>f</kbd> to page-down, so I tried <kbd>ctrl</kbd><kbd>shift</kbd><kbd>f</kbd> instead.

**Comment by andya9** [comment]
@jb Yes, its because it considers the folder you open from to be your project... Maybe we should suggest to make so that, when we launch atom from menu instead of cli, atom will not consider itself in a project.. but i don’t think it would be easy to implement... maybe we could just prevent it from being launched without a path... should I create an issue?

**Comment by jb** [comment]
@andya9: I realized that when I ran find-in-project and it messed up my `~/.atom/` directory...

**Comment by wormyy** [comment]
+1. Just accidentally replaced 'Content-Type' with empty string in the whole project. No way back, I have to revert to last commit.

**Comment by Mystist** [comment]
Suffered the same issue just now...  :sob:

**Comment by xPaw** [comment]
Just occurred to me by mistake. Why is there no confirmation dialog for replace all?

**Comment by PrestonStahley** [comment]
Ouch. +1 for confirmation dialog.

**Comment by DiogoAndre** [comment]
bummer.. something happened to me.

**Comment by hachpai** [comment]
When I opened this issue 6 months ago, i didn't expect all those moments of empathy for other coders.

**Comment by mzampetti** [comment]
+1

**Comment by xPaw** [comment]
Atom 1.0 shipped without fixing this? :angry:

**Comment by dorianfm** [comment]
Would also be good to be able to cancel the operation once started, and to be able to just do a find in project, which was what I was trying to do when Atim replaced the search string with empty space across 38 files

as @xpaw said, this is the kinda thing that shouldn't have made it to 1.0!

**Comment by arthursw** [comment]
+1! `1995 changes in 135 files`

**Comment by kmattimo** [comment]
yow, in atom 1.0 undoing that is broken AND it writes to all the files. That is harsh.

**Comment by chiefGui** [comment]
+1. Please, does someone fix this!

**Comment by chiefGui** [comment]
+1! :cry: 

![image](https://cloud.githubusercontent.com/assets/1330257/8561250/02b41aca-24f9-11e5-94b3-3b9475e2e4c8.png)

**Comment by hermanbergwerf** [comment]
Yes please. I wanted to replace for spaces with one tabs in one file. But because I typed way too fast I actually replaced two spaces by one tab in 52 files 10895 times. I'm going to cry out loud somewhere now.

**Comment by r-murphy** [comment]
+1. 
I like the Sublime approach to this, which is to open all the files and require them to be saved. Along with the corresponding menu opens to save-all and close-all. Although it could probably be improved upon with a warning system or confirmation dialog.

**Comment by iahmad** [comment]
I just replaced all ' to " 
Of course Atom crashed, and my heart as well!

A sad short story!

**Comment by hachpai** [comment]
I've let 30 minutes to myself to find how to just add a confirmation box, but time is over and I've to work... I've never found the definition of replace-all command, seems like auto-generated from the command file or I don't know...

**Comment by thomasjo** [comment]
@hachpai Here it is: [lib/find-view.coffee#L271-L275](https://github.com/atom/find-and-replace/blob/092286a7e9e98cc8bfc831c33c4748eb4b9484b9/lib/find-view.coffee#L271-L275).
The actual command is registered at [line 194](https://github.com/atom/find-and-replace/blob/092286a7e9e98cc8bfc831c33c4748eb4b9484b9/lib/find-view.coffee#L194) in the same file.

**Comment by South-Paw** [comment]
Lost a few hours of work because the find/replace went for the whole file instead of the selection & then I realised there was no undo for it.

It was a fantastic few hours putting that back together.

On a side note; also noticed that general highlighting/selection of text in the editor is rather buggy and the find/replace functions don't always seem to work as expected (such as highlighting a section and then clicking 'in selection' only to have it seemingly deselect your selection)

But anyway. Find/replace needs an undo really badly.

**Comment by MauricioMoraes** [comment]
+1 For this. I was trying to replace within a file and miss-replaced on all files. Lost some time too.

Eclipse has a way to preview the changes that is very useful. You can view the "Future diff" and choose which files you want to modify. It could serve as inspiration for a new tool.

**Comment by RiFi2k** [comment]
" for ' 34406 in 178 files, beat that....

**Comment by park9140** [comment]
@benogle, could we get some attention to this?

**Comment by zaz** [comment]
#446 temporarily disables <kbd>ctrl</kbd> + <kbd>F</kbd>, the hotkey for `project-find`, until a better solution is found.

Am I right in thinking that most people are having this issue because they accidentally used <kbd>ctrl</kbd> + <kbd>F</kbd> instead of <kbd>ctrl</kbd> + <kbd>f</kbd>?

**Comment by RiFi2k** [comment]
Yup I was just moving real fast not paying attention, I would think just having something like if more than 10 files or 100 find and replace actions are going to happen you need to confirm or something to that effect

**Comment by park9140** [comment]
@zaz, for me it's not really that the command is to easy to hit.  It is that there is no catch for the replace action on replace in all files, and it is to easy to trigger the actual replace via keyboard when you are doing a find all.

**Comment by RiFi2k** [comment]
Yeah, agreed. I was sour earlier but now I'm over it. Atom is still my favorite editor I have ever used hands down. Even if I did it 5 more times I still wouldn't switch.
Keep up the good work!

**Comment by r-murphy** [comment]
For me, I meant to hit cmd + F, in order to do a find all across the project, but I did a replace all by accident. 
I entered the find value then tabbed into the replace field to clear it, so the find result wasn't displaying an unrelated replace value. Then I meant to shift+tab back to the find field and press enter to execute. But my right hand was faster than my left and I hit enter when the cursor was still in the replace field.

**Comment by park9140** [comment]
@r-murphy, definitely done it that way at least once.

**Comment by South-Paw** [comment]
as @park9140 said, its not the command that's the issue (for me) its the way that it acts when finding/replacing

**Comment by zaz** [comment]
Well if any of you has an idea of how Atom works internally, a lot of people would appreciate a fix. I don't know how to do much more than just disable hotkeys.

**Comment by danShumway** [comment]
This is a pretty big issue for me as well - I don't think the command is bad, I think the lack of undo for it is just kind of broken in general.  Would there be any internal problems if Atom logged the results of directory-wide actions like this and gave you a way to do some type of global undo?  

I can see this being an issue for other commands from plugins as well.  (Okay, whitespace cleanup on all of my code; oh I didn't commit my actual changes yet and now everything is mixed together)

**Comment by ioquatix** [comment]
Ouch, just got bitten by this.

Could we copy the behaviour from TextMate or something similar? (give you a preview first).

**Comment by soukiassianb** [comment]
+1, Had to manually discard changes on 292 files :hammer:

**Comment by chiefGui** [comment]
@soukiassianb unfortunately, the best way is reverting through GitHub—but this is not always an option. :(

**Comment by bronson** [comment]
If your working dir is dirty, it's not an option.  Just make sure you git commit -a immediately before each project-find-and-replace!  (maybe that's a plugin opportunity for someone...)

**Comment by benogle** [comment]
I'm really sad people are losing work because of this. :(

First, I want to say that just adding an undo command to the way replace currently works seems pretty scary. We'd have to keep track of all the changes in each file, then commit them back to all the replaced files on undo. There seems like all kinds of potential for disaster in doing that. What if a file was changed, moved, or removed etc. between the replace event and the undo? Samesies with a redo.

I've been looking into how other editors handle a project-wide replacement undo. tl;dr: there are 2 ways they handle this: open all the files in the editor, or do nothing.

So what I'm thinking is:

1. Add a confirm dialog
2. Figure out the best thing for step two

## Other Editors

Here is what some other editors do in descending order of undo complexity. I have _not_ found one with an undo that doesn't open all the files in the editor, though I admit that this isn't 100% of the editors out there.

#### Visual Studio

Has a confirm box where you can to change a config setting to enable opening all files in the editor. But once that setting is set, you can undo them all in one action. This is the most advanced undo I found.

![visual-studio-fnr](https://cloud.githubusercontent.com/assets/69169/9019240/4de766f8-379d-11e5-9b2e-bceb7c3042b5.png)

#### Webstorm

They move you into a flow where that opens each file, and you can confirm or skip each replacement.

![webstorm-fnr](https://cloud.githubusercontent.com/assets/69169/9018899/ee268086-3798-11e5-972f-8852c4960f72.png)

#### Sublime

Opens all the files in the editor, requires saving each one.

#### Brackets

Opens all the files in the editor, requires saving each one.

#### Textmate 2

None. The interface is similar to atom:

![textmate-fnr](https://cloud.githubusercontent.com/assets/69169/9018901/ee27829c-3798-11e5-9b3e-48d020a7fefe.png)

#### Notepad++

None. They warn:

![notepadpp-fnr](https://cloud.githubusercontent.com/assets/69169/9018900/ee271d0c-3798-11e5-8e55-47848f7f8cf9.png)

#### Chocolat

None. They warn:

![chocolat-fnr](https://cloud.githubusercontent.com/assets/69169/9018898/ee251e4e-3798-11e5-8732-77b87353a8e1.png)

#### VS Code

Looks like they dont have a replace in files feature.

## What to do beyond a confirm box?

Not sure quite yet. We could follow suit by opening all the files and require saving, but I am worried about the edge cases: blindly opening giant files, or opening a ton of files. So handling those well would be key.

**Comment by ioquatix** [comment]
@benogle I suggest you take a bit more time to review the TextMate find & replace, it has a few critical checks in place which avoid the find and replace disasters which are affecting Atom users.

The most important one IMHO is that you can't click Replace All until you've clicked Find All. This is certainly not the case with Atom.

**Comment by benogle** [comment]
I've opened a PR for step 1 of making the flow for replace-all better: https://github.com/atom/find-and-replace/pull/471

**Comment by South-Paw** [comment]
Worked with Notepad++ a lot before switching to Atom.

I was pretty sure it can undo a find and replace operation...?

**Comment by gigmaps** [comment]
@benogle the best editor I've come across for pretty much all Find & Replace functions is Aptana.
(although there are some rather other serious problems with it that led us here to Atom)

relating specifically to this issue:
- Aptana does not open each file to replace & save
- Aptana has an undo function that groups the last FULL replace and save into the Edit/Undo or ctrl-z
- Aptana undo does not open all files to complete the undo
- Aptana can undo multiple find&replace actions via Edit/Undo or ctrl-z

...if Atom find-and-replace could replicate the Aptana Search functions I'm pretty sure 99% of users would be happy :
- as per find-and-replace there are separate file and project find commands & shortcuts
- results are shown in a separate window that can also be popped out
- results via list or tree view
- multiple parallel project search windows are possible
- search results (including edits) are saved in a history list (number saved is configurable) 
- search input dropdown shows last X number of search terms
- customisable search workspace or working set (including multiple projects in parallel) allows searching of only parts of projects
- expand & contract ALL results with menu button, or file / line with side arrows
- redo button to re-run the search (which removes any edits - like deletes - that have been applied)
- show next/prev match menu button opens file to result location (don't have to select each line
- can delete individual or grouped (eg file) lines from results list (simply using the delete key or menu buttons) so that replace-all is only applied to a tailored result list
- right click on line or file gives option to replace by line, group or project wide
- project wide replacement autosaves all but the currently unsaved files without opening them 
- project wide replacement has preview
- project wide replacement does NOT need a warning as it isn't a single button and has easy undo
- undo (or simply ctrl-z) will undo a project-wide replacement - with multiple return levels (eg if you replaced 3 lines individually, then did a project wide replacement, 4 x ctrl-z has you back where you started)


I highly recommend that you take a look - if we had half the features listed above, this plugin would be 50 times more user friendly.

To summarise: I have Aptana open for project wide find-and-replace, and use Atom for anything _other_ than find-and-replace.

Search Window + right click options:
![screen shot 2016-03-04 at 23 04 00](https://cloud.githubusercontent.com/assets/5176547/13546274/c323ec32-e25d-11e5-935e-7a616b8e3520.png)

When Replace All is Selected (manual selection + easy undo hence no need for pop-up warning)
![screen shot 2016-03-04 at 23 05 42](https://cloud.githubusercontent.com/assets/5176547/13546275/cafea4ba-e25d-11e5-9bb2-0b31d50c5da2.png)

Replace Preview
![screen shot 2016-03-04 at 23 05 52](https://cloud.githubusercontent.com/assets/5176547/13546276/d3a45588-e25d-11e5-8423-f7219ae5d28e.png)

Undo of Replace All (without individual file opening) via Edit or Ctrl-Z
![screen shot 2016-03-04 at 23 06 13](https://cloud.githubusercontent.com/assets/5176547/13546278/d99fcdd2-e25d-11e5-8068-7fe35bb9eed7.png)

**Comment by frozenfung** [comment]
This is a disaster. 😲😲😲

**Comment by cryptogun** [comment]
Oh my. Just replace all \t with spaces. A nightmare. What if my variable contains \t ...
I do double clicked one file and enter that file, and then replace. I wanna do the replacement one file at a time carefully, but all files in project are replaced. Unable to undo.
Now I can only use `git checkout HEAD -- my-file.txt` to reset a file. But some uncommited changes are also gone.

**Comment by shei25420** [comment]
you guys just gave me heart was about to cry when i saw replaced 12234 in 76 files...but atleast am not alone on this one...cmon atom developers u'll one day cause a suicide if you don't try to fix this issue a confirm box that tell you the changes will be applied on all files is enough man

**Comment by bekahmcdonald** [comment]
Ugh, this just happened to me as well, for the second time. Confirm box, no automatic save, or at the very least an undo please!

**Comment by 50Wliu** [comment]
Are people who are experiencing this not seeing the following dialog when pressing the Replace All button?
![find-and-replace-confirm-dialog](https://user-images.githubusercontent.com/2766036/32700177-c0541ce8-c7c1-11e7-8da1-c20658d698f3.png)
Or are you asking for a different kind of "confirm box"?

**Comment by bekahmcdonald** [comment]
Nope, @50Wliu, definitely not seeing that dialog box!

**Comment by 50Wliu** [comment]
@bekahmcdonald Can you please open a new issue regarding the confirm dialog not appearing and fill out the issue template?  That would really help 🙇

**Comment by mattsears18** [comment]
I just replaced every comma in my project and I have no way of undoing it. FML.

**Comment by chriswayg** [comment]
I was really surprised, that there is no undo. The confirm box should at least warn, that there is no undo for Replace All, like:

## :exclamation: This action cannot be undone :exclamation:

**Comment by MangoSmoothie101010** [comment]
This MUST be undoable.

I see, there is no bounty yet awarded for this.

Now there is a small bounty of 5$ for starters:
https://www.bountysource.com/issues/7115106-add-undo-to-replace-all-in-project

**Comment by gagecarto** [comment]
noooooooo. ouch.. this one hurts.. i can't believe there's an action in an editor that won't respond to a simple undo.. bummer

**Comment by walteryu** [comment]
Same issue here; I wanted to find/replace in 1 file, then it crawled all files in my entire directory and made hundreds of edits. Fortunately, I was unable to undo the change by find/replace the previous value back (though this may not always be the case) so best to have Git/version control ready.

**Comment by ptoche** [comment]
> Are people who are experiencing this not seeing the following dialog when pressing the Replace All button?
> ![find-and-replace-confirm-dialog](https://user-images.githubusercontent.com/2766036/32700177-c0541ce8-c7c1-11e7-8da1-c20658d698f3.png)
> Or are you asking for a different kind of "confirm box"?

No confirm box here.

**Comment by giampaolo44** [comment]
Lol. Such a critical issue has been open for 6 years? I'd be curious to know how many users it has lost to Atom, and I wish I had the time and skill to look into it.
Intuitively it does not sound like an unsolvable problem. Anyone tried to make it right so far? What's the issue there?

**Comment by finnbear** [comment]
Just replaced all instances of "min(" with "max(". I don't know if the math in my project will ever work 100% properly again :confounded:

**Comment by rafapolo** [comment]
Giving up of Atom after this one, knowing it is hurting people for the past 8 years. 
I've just removed all regex for numbers in 263 files that I was git-less prototyping for the past year.
A weekly backup can't help much. Project is dead.
Were all developers bought by Microsoft and forced to make this beloved IDE obsolete in name of VSCode?
